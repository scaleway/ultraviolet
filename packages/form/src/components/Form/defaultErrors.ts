import type { RequiredErrors } from '../../types'

export const defaultErrors: RequiredErrors = {
  deps: () => '',
  disabled: () => '',
  max: () => '',
  maxDate: () => '',
  maxLength: () => '',
  min: () => '',
  minLength: () => '',
  onBlur: () => '',
  onChange: () => '',
  pattern: () => '',
  required: () => '',
  setValueAs: () => '',
  shouldUnregister: () => '',
  validate: () => '',
  value: () => '',
  valueAsDate: () => '',
  valueAsNumber: () => '',
}
