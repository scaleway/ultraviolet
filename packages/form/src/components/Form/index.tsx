'use client'

import styled from '@emotion/styled'
import type { ReactNode } from 'react'
import type { FieldValues, UseFormReturn } from 'react-hook-form'
import { FormProvider } from 'react-hook-form'
import { ErrorProvider } from '../../providers'
import type { FormErrors } from '../../types'
import { defaultErrors } from './defaultErrors'

type OnSubmitReturn = string | null | undefined | void

type FormProps<TFieldValues extends FieldValues> = {
  children?: ReactNode
  errors: FormErrors
  name?: string
  onSubmit: (data: TFieldValues) => Promise<OnSubmitReturn> | OnSubmitReturn
  methods: UseFormReturn<TFieldValues>
}

const StyledForm = styled.form`
  display: contents;
`

export const Form = <TFieldValues extends FieldValues>({
  children,
  methods,
  errors,
  onSubmit,
  name,
}: FormProps<TFieldValues>) => {
  const handleSubmit = methods.handleSubmit(async values => {
    const result = await onSubmit(values)

    if (result) {
      methods.setError('root.submit', {
        message: result,
        type: 'custom',
      })
    }
  })

  return (
    <FormProvider {...methods}>
      <ErrorProvider errors={{ ...defaultErrors, ...errors }}>
        <StyledForm
          name={name}
          noValidate
          onSubmit={async e => {
            e.preventDefault()
            e.stopPropagation()
            await handleSubmit(e)
          }}
        >
          {children}
        </StyledForm>
      </ErrorProvider>
    </FormProvider>
  )
}
