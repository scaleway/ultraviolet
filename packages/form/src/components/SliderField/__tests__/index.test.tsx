import { act, fireEvent, renderHook, screen } from '@testing-library/react'
import { userEvent } from '@testing-library/user-event'
import { mockFormErrors, renderWithForm, renderWithTheme } from '@utils/test'
import { useForm } from 'react-hook-form'
import { describe, expect, test, vi } from 'vitest'
import { SliderField, Submit } from '../..'
import { Form } from '../../Form'

const options = [
  { label: '1Mb', value: 1 },
  { label: '3Mb', value: 3 },
  { label: '5Mb', value: 5 },
  { label: '10Mb', value: 10 },
  { label: '50Mb', value: 50 },
  { label: '100Mb', value: 100 },
  { label: '500Mb', value: 500 },
]

describe('SliderField', () => {
  test('should render correctly', () => {
    const { asFragment } = renderWithForm(<SliderField name="test" value={0} />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('should render correctly with possible values', () => {
    const { asFragment } = renderWithForm(
      <SliderField name="test" options={options} />,
    )
    expect(asFragment()).toMatchSnapshot()
  })

  test('should render correctly with possible values and double', () => {
    const { asFragment } = renderWithForm(
      <SliderField double name="test" options={options} />,
    )
    expect(asFragment()).toMatchSnapshot()
  })

  test('should render correctly disabled', () => {
    const { asFragment } = renderWithForm(
      <SliderField aria-label="Slider Input" disabled name="test" value={10} />,
    )
    const input = screen.getAllByLabelText('Slider Input')[1]
    expect(input).toBeDisabled()

    expect(asFragment()).toMatchSnapshot()
  })

  test('should trigger events correctly', () => {
    const onFocus = vi.fn(() => {})
    const onBlur = vi.fn(() => {})

    const { asFragment } = renderWithForm(
      <SliderField
        input
        label="Test"
        max={10}
        min={0}
        name="test"
        onBlur={onBlur}
        onFocus={onFocus}
        required
      />,
    )
    const input = screen.getByRole('slider', { hidden: true })
    act(() => input.focus())
    expect(onFocus).toBeCalledTimes(1)
    act(() => input.blur())
    expect(onBlur).toBeCalledTimes(1)
    expect(asFragment()).toMatchSnapshot()
  })

  test('should work fine with form setValue', async () => {
    const onSubmit = vi.fn()
    const onBlur = vi.fn(() => {})
    const { result } = renderHook(() =>
      useForm<{ test: number | null }>({
        defaultValues: {
          test: 10,
        },
        mode: 'onChange',
      }),
    )

    renderWithTheme(
      <Form
        errors={mockFormErrors}
        methods={result.current}
        onSubmit={value => onSubmit(value)}
      >
        <SliderField
          input
          label="Test"
          max={10}
          min={0}
          name="test"
          onBlur={onBlur}
          required
        />
        <Submit>Submit</Submit>
      </Form>,
    )

    const slider = screen.getByRole<HTMLInputElement>('slider')
    const submit = screen.getByText('Submit')
    expect(slider).toHaveValue('10')
    await userEvent.click(submit)
    expect(onSubmit).toHaveBeenCalledWith({
      test: 10,
    })

    const input = screen.getByTestId<HTMLInputElement>('slider-input')
    await userEvent.type(input, '5')
    await userEvent.tab()
    expect(slider.value).toBe('10')
  })

  test('should work correctly with possibleValues', () => {
    const { result } = renderHook(() => useForm({ mode: 'onChange' }))

    renderWithTheme(
      <Form
        errors={mockFormErrors}
        methods={result.current}
        onSubmit={() => {}}
      >
        <SliderField label="Test" name="test" options={options} />
      </Form>,
    )

    const input = screen.getByRole('slider', { hidden: true })

    fireEvent.change(input, { target: { value: '2' } })
    expect(input).toHaveValue('2')
    expect(result.current.getValues('test')).toBe(5)

    fireEvent.change(input, { target: { value: '5' } })
    expect(input).toHaveValue('5')
    expect(result.current.getValues('test')).toBe(100)
  })

  test('should work correctly with possibleValues and double', () => {
    const { result } = renderHook(() => useForm({ mode: 'onChange' }))

    renderWithTheme(
      <Form
        errors={mockFormErrors}
        methods={result.current}
        onSubmit={() => {}}
      >
        <SliderField double label="Test" name="test" options={options} />
      </Form>,
    )

    const input1 = screen.getAllByRole('slider', { hidden: true })[0]
    const input2 = screen.getAllByRole('slider', { hidden: true })[1]

    fireEvent.change(input1, { target: { value: '2' } })
    expect(input1).toHaveValue('2')
    expect(result.current.getValues('test')).toStrictEqual([5, 500])

    fireEvent.change(input2, { target: { value: '5' } })
    expect(input2).toHaveValue('5')
    expect(result.current.getValues('test')).toStrictEqual([5, 100])
  })
})
