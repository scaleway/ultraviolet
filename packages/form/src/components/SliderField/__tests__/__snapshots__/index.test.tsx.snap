// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sliderField > should render correctly 1`] = `
<DocumentFragment>
  <div
    data-testid="testing"
  >
    <form
      novalidate=""
      style="display: contents;"
    >
      <div
        class="uv_toi52u0 uv_toi52u31 uv_toi52u2d uv_toi52u7d uv_toi52u7 uv_toi52u5d"
        data-options="false"
      >
        <div
          class="uv_toi52u0 uv_toi52u31 uv_toi52u2d uv_toi52u7d uv_toi52u7 uv_toi52u6d"
        >
          <div
            class="uv_toi52u0 uv_toi52u31 uv_toi52u2j uv_toi52u7d uv_toi52u5v"
          >
            <span
              class="uv_a5ifr1h uv_a5ifr1k uv_a5ifr1i uv_m4c9ow0 uv_m4c9ow2 uv_m4c9ow4 uv_m4c9ow6 uv_m4c9ow8 uv_m4c9owb uv_m4c9owi uv_m4c9own uv_m4c9ow1b uv_m4c9ow20 uv_m4c9ow3c"
              data-testid="slider-value"
              style="--uv_qabug40: right;"
            >
              0
            </span>
          </div>
          <div
            class="uv_toi52u0 uv_toi52u31 uv_toi52u2d uv_toi52u7d uv_toi52u7 uv_toi52u5j"
            style="--uv_4k0ekn0: 100%;"
          >
            <input
              aria-label="test"
              class="uv_a5ifr1p uv_a5ifr1t uv_a5ifr1b uv_a5ifr1d uv_a5ifr1f"
              data-direction="column"
              data-error="false"
              id="_r_0_"
              max="100"
              min="0"
              name="test"
              step="1"
              style="--uv_a5ifr12: calc(0% - 8px); --uv_a5ifr11: #ffffff; background-size: 0% 100%;"
              tabindex="0"
              type="range"
              value="0"
            />
          </div>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`sliderField > should render correctly disabled 1`] = `
<DocumentFragment>
  <div
    data-testid="testing"
  >
    <form
      novalidate=""
      style="display: contents;"
    >
      <div
        aria-label="Slider Input"
        class="uv_toi52u0 uv_toi52u31 uv_toi52u2d uv_toi52u7d uv_toi52u7 uv_toi52u5d"
        data-options="false"
      >
        <div
          class="uv_toi52u0 uv_toi52u31 uv_toi52u2d uv_toi52u7d uv_toi52u7 uv_toi52u6d"
        >
          <div
            class="uv_toi52u0 uv_toi52u31 uv_toi52u2j uv_toi52u7d uv_toi52u5v"
          >
            <span
              class="uv_a5ifr1h uv_a5ifr1k uv_a5ifr1i uv_m4c9ow0 uv_m4c9ow2 uv_m4c9ow4 uv_m4c9ow6 uv_m4c9ow8 uv_m4c9owb uv_m4c9owi uv_m4c9own uv_m4c9ow1b uv_m4c9ow20 uv_m4c9ow3c"
              data-testid="slider-value"
              style="--uv_qabug40: right;"
            >
              0
            </span>
          </div>
          <div
            class="uv_toi52u0 uv_toi52u31 uv_toi52u2d uv_toi52u7d uv_toi52u7 uv_toi52u5j"
            style="--uv_4k0ekn0: 100%;"
          >
            <input
              aria-disabled="true"
              aria-label="Slider Input"
              class="uv_a5ifr1p uv_a5ifr1t uv_a5ifr1r uv_a5ifr1b uv_a5ifr1c uv_a5ifr1f"
              data-direction="column"
              data-error="false"
              disabled=""
              id="_r_q_"
              max="100"
              min="0"
              name="test"
              step="1"
              style="--uv_a5ifr12: calc(0% - 8px); --uv_a5ifr11: #ffffff; background-size: 0% 100%;"
              tabindex="0"
              type="range"
              value="0"
            />
          </div>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`sliderField > should render correctly with possible values 1`] = `
<DocumentFragment>
  <div
    data-testid="testing"
  >
    <form
      novalidate=""
      style="display: contents;"
    >
      <div
        class="uv_toi52u0 uv_toi52u31 uv_toi52u2d uv_toi52u7d uv_toi52uj uv_toi52u5d"
        data-options="true"
      >
        <div
          class="uv_toi52u0 uv_toi52u31 uv_toi52u2d uv_toi52u7d uv_toi52u7 uv_toi52u6d"
        >
          <div
            class="uv_toi52u0 uv_toi52u31 uv_toi52u2j uv_toi52u7d uv_toi52u5v"
          >
            <span
              class="uv_a5ifr1h uv_a5ifr1k uv_a5ifr1i uv_m4c9ow0 uv_m4c9ow2 uv_m4c9ow4 uv_m4c9ow6 uv_m4c9ow8 uv_m4c9owb uv_m4c9owi uv_m4c9own uv_m4c9ow1b uv_m4c9ow20 uv_m4c9ow3c"
              data-testid="slider-value"
              style="--uv_qabug40: right;"
            >
              1
            </span>
          </div>
          <div
            class="uv_toi52u0 uv_toi52u31 uv_toi52u2d uv_toi52u7d uv_toi52u7 uv_toi52u5j"
            style="--uv_4k0ekn0: 100%;"
          >
            <input
              aria-label="test"
              class="uv_a5ifr1p uv_a5ifr1t uv_a5ifr1b uv_a5ifr1d uv_a5ifr1f"
              data-direction="column"
              data-error="false"
              id="_r_3_"
              max="6"
              min="0"
              name="test"
              step="1"
              style="--uv_a5ifr12: calc(0% - 8px); --uv_a5ifr11: #ffffff; background-size: 0% 100%;"
              tabindex="0"
              type="range"
              value="0"
            />
            <datalist
              class="uv_a5ifr16"
              data-double="false"
            >
              <span
                class="uv_a5ifr18 uv_a5ifr19"
                data-element-left="true"
                data-element-right="false"
                data-value="0"
                style="--uv_a5ifr13: 0%;"
              >
                <p
                  class="uv_m4c9ow0 uv_m4c9ow2 uv_m4c9ow4 uv_m4c9ow6 uv_m4c9ow8 uv_m4c9owb uv_m4c9owi uv_m4c9ows uv_m4c9ow1b uv_m4c9ow20 uv_m4c9ow3c"
                >
                  1Mb
                </p>
              </span>
              <span
                class="uv_a5ifr18"
                data-element-left="false"
                data-element-right="false"
                data-value="1"
                style="--uv_a5ifr13: 16.666666666666668%;"
              >
                <p
                  class="uv_m4c9ow0 uv_m4c9ow2 uv_m4c9ow4 uv_m4c9ow6 uv_m4c9ow8 uv_m4c9owb uv_m4c9owi uv_m4c9ows uv_m4c9ow1b uv_m4c9ow20 uv_m4c9ow3c"
                >
                  3Mb
                </p>
              </span>
              <span
                class="uv_a5ifr18"
                data-element-left="false"
                data-element-right="false"
                data-value="2"
                style="--uv_a5ifr13: 33.333333333333336%;"
              >
                <p
                  class="uv_m4c9ow0 uv_m4c9ow2 uv_m4c9ow4 uv_m4c9ow6 uv_m4c9ow8 uv_m4c9owb uv_m4c9owi uv_m4c9ows uv_m4c9ow1b uv_m4c9ow20 uv_m4c9ow3c"
                >
                  5Mb
                </p>
              </span>
              <span
                class="uv_a5ifr18"
                data-element-left="false"
                data-element-right="false"
                data-value="3"
                style="--uv_a5ifr13: 50%;"
              >
                <p
                  class="uv_m4c9ow0 uv_m4c9ow2 uv_m4c9ow4 uv_m4c9ow6 uv_m4c9ow8 uv_m4c9owb uv_m4c9owi uv_m4c9ows uv_m4c9ow1b uv_m4c9ow20 uv_m4c9ow3c"
                >
                  10Mb
                </p>
              </span>
              <span
                class="uv_a5ifr18"
                data-element-left="false"
                data-element-right="false"
                data-value="4"
                style="--uv_a5ifr13: 66.66666666666667%;"
              >
                <p
                  class="uv_m4c9ow0 uv_m4c9ow2 uv_m4c9ow4 uv_m4c9ow6 uv_m4c9ow8 uv_m4c9owb uv_m4c9owi uv_m4c9ows uv_m4c9ow1b uv_m4c9ow20 uv_m4c9ow3c"
                >
                  50Mb
                </p>
              </span>
              <span
                class="uv_a5ifr18"
                data-element-left="false"
                data-element-right="false"
                data-value="5"
                style="--uv_a5ifr13: 83.33333333333333%;"
              >
                <p
                  class="uv_m4c9ow0 uv_m4c9ow2 uv_m4c9ow4 uv_m4c9ow6 uv_m4c9ow8 uv_m4c9owb uv_m4c9owi uv_m4c9ows uv_m4c9ow1b uv_m4c9ow20 uv_m4c9ow3c"
                >
                  100Mb
                </p>
              </span>
              <span
                class="uv_a5ifr18 uv_a5ifr1a"
                data-element-left="false"
                data-element-right="true"
                data-value="6"
                style="--uv_a5ifr13: 100%;"
              >
                <p
                  class="uv_m4c9ow0 uv_m4c9ow2 uv_m4c9ow4 uv_m4c9ow6 uv_m4c9ow8 uv_m4c9owb uv_m4c9owi uv_m4c9ows uv_m4c9ow1b uv_m4c9ow20 uv_m4c9ow3c"
                >
                  500Mb
                </p>
              </span>
            </datalist>
          </div>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`sliderField > should render correctly with possible values and double 1`] = `
<DocumentFragment>
  <div
    data-testid="testing"
  >
    <form
      novalidate=""
      style="display: contents;"
    >
      <div
        class="uv_a5ifr14 uv_toi52u0 uv_toi52u31 uv_toi52u2d uv_toi52u7d uv_toi52uj uv_toi52u5d"
        data-double="true"
        data-options="true"
      >
        <div
          class="uv_toi52u0 uv_toi52u31 uv_toi52u2d uv_toi52u7d uv_toi52u7 uv_toi52u6d"
        >
          <div
            class="uv_toi52u0 uv_toi52u31 uv_toi52u2d uv_toi52u7d uv_toi52u7 uv_toi52u5d"
            style="--uv_4k0ekn0: 100%;"
          >
            <div
              class="uv_toi52u0 uv_toi52u31 uv_toi52u2j uv_toi52u7d uv_toi52u6p"
            >
              <span
                class="uv_a5ifr1h uv_a5ifr1j uv_m4c9ow0 uv_m4c9ow2 uv_m4c9ow4 uv_m4c9ow6 uv_m4c9ow8 uv_m4c9owb uv_m4c9owi uv_m4c9own uv_m4c9ow1b uv_m4c9ow20 uv_m4c9ow3c"
                data-testid="value-left"
                style="--uv_qabug40: right;"
              >
                1
              </span>
              <span
                class="uv_a5ifr1h uv_a5ifr1j uv_m4c9ow0 uv_m4c9ow2 uv_m4c9ow4 uv_m4c9ow6 uv_m4c9ow8 uv_m4c9owb uv_m4c9owi uv_m4c9own uv_m4c9ow1b uv_m4c9ow20 uv_m4c9ow3c"
                data-testid="value-right"
                style="--uv_qabug40: right;"
              >
                500
              </span>
            </div>
            <div
              class="uv_a5ifr1v"
            >
              <div
                class="uv_a5ifr1w"
              >
                <div
                  aria-disabled="false"
                  class="uv_a5ifr1x"
                  data-error="false"
                  style="left: 0%; right: 0%;"
                />
              </div>
              <input
                aria-label="test"
                class="uv_a5ifr1l uv_a5ifr1n uv_a5ifr1b uv_a5ifr1d uv_a5ifr1e uv_a5ifr1g"
                data-direction="column"
                data-testid="handle-left"
                id="_r_d_"
                max="6"
                min="0"
                name="test"
                step="1"
                style="--uv_a5ifr12: calc(0% - 8px; --uv_a5ifr11: #ffffff;"
                type="range"
                value="0"
              />
              <input
                aria-label="test"
                class="uv_a5ifr1l uv_a5ifr1n uv_a5ifr1b uv_a5ifr1d uv_a5ifr1e uv_a5ifr1g"
                data-direction="column"
                data-testid="handle-right"
                id="_r_d_"
                max="6"
                min="0"
                name="test"
                step="1"
                style="--uv_a5ifr12: calc(100% - 8px; --uv_a5ifr11: #ffffff;"
                type="range"
                value="6"
              />
              <datalist
                class="uv_a5ifr16 uv_a5ifr17"
                data-double="true"
              >
                <span
                  class="uv_a5ifr18 uv_a5ifr19"
                  data-element-left="true"
                  data-element-right="false"
                  data-value="0"
                  style="--uv_a5ifr13: 0%;"
                >
                  <p
                    class="uv_m4c9ow0 uv_m4c9ow2 uv_m4c9ow4 uv_m4c9ow6 uv_m4c9ow8 uv_m4c9ow9 uv_m4c9owi uv_m4c9oww uv_m4c9ow1b uv_m4c9ow1c uv_m4c9ow3c"
                  >
                    1Mb
                  </p>
                </span>
                <span
                  class="uv_a5ifr18"
                  data-element-left="false"
                  data-element-right="false"
                  data-value="1"
                  style="--uv_a5ifr13: 16.666666666666668%;"
                >
                  <p
                    class="uv_m4c9ow0 uv_m4c9ow2 uv_m4c9ow4 uv_m4c9ow6 uv_m4c9ow8 uv_m4c9owb uv_m4c9owi uv_m4c9ows uv_m4c9ow1b uv_m4c9ow20 uv_m4c9ow3c"
                  >
                    3Mb
                  </p>
                </span>
                <span
                  class="uv_a5ifr18"
                  data-element-left="false"
                  data-element-right="false"
                  data-value="2"
                  style="--uv_a5ifr13: 33.333333333333336%;"
                >
                  <p
                    class="uv_m4c9ow0 uv_m4c9ow2 uv_m4c9ow4 uv_m4c9ow6 uv_m4c9ow8 uv_m4c9owb uv_m4c9owi uv_m4c9ows uv_m4c9ow1b uv_m4c9ow20 uv_m4c9ow3c"
                  >
                    5Mb
                  </p>
                </span>
                <span
                  class="uv_a5ifr18"
                  data-element-left="false"
                  data-element-right="false"
                  data-value="3"
                  style="--uv_a5ifr13: 50%;"
                >
                  <p
                    class="uv_m4c9ow0 uv_m4c9ow2 uv_m4c9ow4 uv_m4c9ow6 uv_m4c9ow8 uv_m4c9owb uv_m4c9owi uv_m4c9ows uv_m4c9ow1b uv_m4c9ow20 uv_m4c9ow3c"
                  >
                    10Mb
                  </p>
                </span>
                <span
                  class="uv_a5ifr18"
                  data-element-left="false"
                  data-element-right="false"
                  data-value="4"
                  style="--uv_a5ifr13: 66.66666666666667%;"
                >
                  <p
                    class="uv_m4c9ow0 uv_m4c9ow2 uv_m4c9ow4 uv_m4c9ow6 uv_m4c9ow8 uv_m4c9owb uv_m4c9owi uv_m4c9ows uv_m4c9ow1b uv_m4c9ow20 uv_m4c9ow3c"
                  >
                    50Mb
                  </p>
                </span>
                <span
                  class="uv_a5ifr18"
                  data-element-left="false"
                  data-element-right="false"
                  data-value="5"
                  style="--uv_a5ifr13: 83.33333333333333%;"
                >
                  <p
                    class="uv_m4c9ow0 uv_m4c9ow2 uv_m4c9ow4 uv_m4c9ow6 uv_m4c9ow8 uv_m4c9owb uv_m4c9owi uv_m4c9ows uv_m4c9ow1b uv_m4c9ow20 uv_m4c9ow3c"
                  >
                    100Mb
                  </p>
                </span>
                <span
                  class="uv_a5ifr18 uv_a5ifr1a"
                  data-element-left="false"
                  data-element-right="true"
                  data-value="6"
                  style="--uv_a5ifr13: 100%;"
                >
                  <p
                    class="uv_m4c9ow0 uv_m4c9ow2 uv_m4c9ow4 uv_m4c9ow6 uv_m4c9ow8 uv_m4c9ow9 uv_m4c9owi uv_m4c9oww uv_m4c9ow1b uv_m4c9ow1c uv_m4c9ow3c"
                  >
                    500Mb
                  </p>
                </span>
              </datalist>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`sliderField > should trigger events correctly 1`] = `
<DocumentFragment>
  <div
    data-testid="testing"
  >
    <form
      novalidate=""
      style="display: contents;"
    >
      <div
        class="uv_toi52u0 uv_toi52u31 uv_toi52u2d uv_toi52u7d uv_toi52u7 uv_toi52u5d"
        data-options="false"
      >
        <div
          class="uv_toi52u0 uv_toi52u31 uv_toi52u2d uv_toi52u7d uv_toi52u7 uv_toi52u6d"
        >
          <div
            class="uv_toi52u0 uv_toi52u3d uv_toi52u2j uv_toi52u7d uv_toi52u6p"
          >
            <div
              class="uv_toi52u0 uv_toi52u3j uv_toi52u2j uv_toi52u7d uv_toi52u1v uv_toi52u5d"
            >
              <label
                class="uv_17td3g91 uv_m4c9ow0 uv_m4c9ow2 uv_m4c9ow4 uv_m4c9ow6 uv_m4c9ow8 uv_m4c9owb uv_m4c9owi uv_m4c9owq uv_m4c9ow1b uv_m4c9ow20 uv_m4c9ow3c"
                for="_r_t_"
              >
                Test
              </label>
              <span
                class="uv_m4c9ow0 uv_m4c9ow2 uv_m4c9ow4 uv_m4c9ow6 uv_m4c9ow8 uv_m4c9owd uv_m4c9owi uv_m4c9owq uv_m4c9ow1b uv_m4c9ow1k uv_m4c9ow3c"
              >
                *
              </span>
            </div>
            <div
              class="uv_a5ifr110 uv_toi52u0 uv_toi52u31 uv_toi52u2d uv_toi52u7d uv_toi52u1v uv_toi52u5d"
            >
              <div>
                <div
                  class="uv_1ce7gb6h uv_1ce7gb6n uv_1ce7gb6i"
                  data-controls="false"
                  data-disabled="false"
                  data-error="false"
                  data-size="small"
                  data-success="false"
                  data-unit="false"
                >
                  <div
                    class="uv_1ce7gb63 uv_1ce7gb64 uv_x6hyh50 uv_x6hyh52p uv_x6hyh561"
                    style="--uv_3mg8xe0: 1fr auto; --uv_3mg8xe1: 1fr auto; --uv_3mg8xe2: 1fr auto; --uv_3mg8xe3: 1fr auto; --uv_3mg8xe4: 1fr auto; --uv_3mg8xe5: 1fr auto;"
                  >
                    <input
                      aria-label="input"
                      class="uv_1ce7gb6b uv_1ce7gb6d uv_1ce7gb6g"
                      data-testid="slider-input"
                      id="_r_10_"
                      max="10"
                      min="0"
                      placeholder=""
                      step="1"
                      type="number"
                      value="0"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="uv_toi52u0 uv_toi52u31 uv_toi52u2j uv_toi52u7d uv_toi52u5v"
          />
          <div
            class="uv_toi52u0 uv_toi52u31 uv_toi52u2d uv_toi52u7d uv_toi52u7 uv_toi52u5j"
            style="--uv_4k0ekn0: 100%;"
          >
            <input
              aria-label="test"
              class="uv_a5ifr1p uv_a5ifr1t uv_a5ifr1b uv_a5ifr1d uv_a5ifr1f"
              data-direction="column"
              data-error="false"
              id="_r_t_"
              max="10"
              min="0"
              name="test"
              step="1"
              style="--uv_a5ifr12: calc(0% - 8px); --uv_a5ifr11: #ffffff; background-size: 0% 100%;"
              tabindex="0"
              type="range"
              value="0"
            />
          </div>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
